ocaml/xapi/xapi_pool.ml

```diff
@@ -3717,3 +3748,44 @@ let get_guest_secureboot_readiness ~__context ~self:_ =
       `ready_no_dbx
   | _, _, _, _ ->
       `not_ready
+
+let unpack_bundle ifd =
+  let result =
+    Tar_ext.unpack_tar_file
+      ~dir:!Xapi_globs.bundle_repository_dir
+      ~ifd
+      ~max_size_limit:!Xapi_globs.bundle_max_size_limit
+  in
+  match result with
+  | Ok () ->
+      ()
+  | Error e ->
+      raise
+        Api_errors.(
+          Server_error
+            (bundle_unpacked_failed, [Tar_ext.unpack_error_to_string e])
+        )
+
+let put_bundle_handler (req : Request.t) s _ =
```

Comment:
May I know how will the `task_id` in the request be handled?

Reply:
The client (XC) will create a task before uploading the bundle. The uploading request will contain `task_id` in `req`. Then the `task_id` will be handled in function `Xapi_http.with_context` and forward function `Server_helpers.exec_with_forwarded_task`.

Reply:
Discussed with XC team, currently XC doesn't use the progress of task. It uses self-implemented progress bar. There may be 2 reasons:
1 XC will trace the task after uploading is finished.
2 The task defined in xapi doesn't contain the actual bytes uploaded.